src/main/java/edu/cornell/kfs/fp/batch/service/impl/AuxiliaryVoucherDocumentGenerator.java

```diff
@@ -0,0 +1,234 @@
+        return accountingLine;
+    }
+
+    protected <A extends AccountingLine> void validateAndPopulateAccountingLineWithDebitCreditAmounts(
+            A accountingLine, AccountingXmlDocumentAccountingLine xmlLine) {
+        if (hasAmountBeenSpecified(xmlLine.getAmount())) {
+            throw new ValidationException(
+                    "AV accounting line cannot use the 'amount' element to specify an amount; use the relevant debit/credit amount element instead");
+        }
+        
+        KualiDecimal debitAmount = xmlLine.getDebitAmount();
+        KualiDecimal creditAmount = xmlLine.getCreditAmount();
+        if (hasAmountBeenSpecified(debitAmount)) {
+            if (hasAmountBeenSpecified(creditAmount)) {
+                throw new ValidationException("AV accounting line cannot specify both a debit amount and a credit amount");
+            }
+            accountingLine.setDebitCreditCode(KFSConstants.GL_DEBIT_CODE);
+            accountingLine.setAmount(debitAmount);
+        } else if (hasAmountBeenSpecified(creditAmount)) {
+            accountingLine.setDebitCreditCode(KFSConstants.GL_CREDIT_CODE);
+            accountingLine.setAmount(creditAmount);
+        } else {
+            throw new ValidationException("AV accounting line did not specify a debit or credit amount");
+        }
+    }
+
+    protected boolean hasAmountBeenSpecified(KualiDecimal amount) {
+        return amount != null && !KualiDecimal.ZERO.equals(amount);
+    }
+
+    @Override
+    protected void populateCustomAccountingDocumentData(AuxiliaryVoucherDocument document, AccountingXmlDocumentEntry documentEntry) {
+        DateTime documentCreateDate = getDocumentCreateDate(document);
+        Date documentCreateDateWithoutTime = new Date(documentCreateDate.getMillis());
+        
+        AccountingPeriod period = findEligibleAccountingPeriod(document, documentEntry);
+        document.setAccountingPeriod(period);
+        
+        String avTypeCode = validateAndGetAuxiliaryVoucherType(documentEntry);
+        document.setTypeCode(avTypeCode);
+        
+        Optional<Date> reversalDate = validateAndGetReversalDateIfApplicable(
+                documentEntry, period, avTypeCode, documentCreateDateWithoutTime);
+        if (reversalDate.isPresent()) {
+            document.setReversalDate(reversalDate.get());
+        }
+    }
+
+    protected DateTime getDocumentCreateDate(AuxiliaryVoucherDocument document) {
+        WorkflowDocument workflowDocument = document.getDocumentHeader().getWorkflowDocument();
+        return workflowDocument.getDateCreated();
+    }
+
+    /*
+     * NOTE: Much of this period-searching logic has been duplicated and modified from AuxiliaryVoucherForm.
+     * If KualiCo updates the related base code to be more micro-test-friendly, then we should update
+     * the code below accordingly.
+     */
+    protected AccountingPeriod findEligibleAccountingPeriod(
```

Comment:
Can this be simplified if the XML document takes in a numberic accounting period and let the document validate if the accounting period is valid?

Reply:
After further analysis, it looks like the AV document validation rules do contain sufficient checks for ensuring that the accounting period is eligible for use. I will simplify this piece further to just check that the referenced accounting period actually exists.

As for searching by year-and-period-code instead of by period name, I will check with the functionals to see if they are fine with such a setup.